src/closed.jl

```diff
@@ -0,0 +1,44 @@
+"""
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are closed i.e. the range is [left, right]
+"""
+immutable ClosedSet{T}
+	left::T
+	right::T
+end
+
+function ClosedSet(left, right)
+	(I_left, I_right) = promote(left, right)
+	ClosedSet(I_left, I_right)
+end
+
+show(io::IO, I::ClosedSet) = print(io, "[", I.left, ",", I.right, "]")
+
+length(I::ClosedSet) = I.right - I.left + 1
+
+in(v, I::ClosedSet) = I.left <= v <= I.right
+
+isempty(A::ClosedSet) = A.left > A.right
+
+isequal(A::ClosedSet, B::ClosedSet) = (A.left == B.left && A.right == B.right) || isempty(A) && isempty(B)
+
+==(A::ClosedSet, B::ClosedSet) = isequal(A, B)
+
+function intersect(A::ClosedSet, B::ClosedSet)
+	left = max(A.left, B.left)
+	right = min(A.right, B.right)
+	left, right = promote(left, right)
```

Comment:
You don't need this line, because the constructor will take care of that. Moreover, `left` and `right` should already be of the same type as each other, even if they differ from the type of `A` and/or `B`, because `min` and `max` promote.